In [1]:
import pandas as pd
import warnings
warnings.filterwarnings ('ignore')

In [2]:
df_features=pd.read_csv('features.csv')

In [3]:
df_features.head(2)

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
0,1,2010-02-05,42.31,2.572,NaN,NaN,NaN,NaN,NaN,211.096358,8.106,False
1,1,2010-02-12,38.51,2.548,NaN,NaN,NaN,NaN,NaN,211.242170,8.106,True


In [4]:
df_stores=pd.read_csv('stores.csv')

In [5]:
df_stores.head(2)

,Store,Type,Size
0,1,A,151315
1,2,A,202307


In [6]:
df_train=pd.read_csv('train.csv')

In [7]:
df_train.head(2)

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,2010-02-05,24924.50,False
1,1,1,2010-02-12,46039.49,True


# data cleaning # df_features

In [8]:
df_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         8190 non-null   int64  
 1   Date          8190 non-null   object 
 2   Temperature   8190 non-null   float64
 3   Fuel_Price    8190 non-null   float64
 4   MarkDown1     4032 non-null   float64
 5   MarkDown2     2921 non-null   float64
 6   MarkDown3     3613 non-null   float64
 7   MarkDown4     3464 non-null   float64
 8   MarkDown5     4050 non-null   float64
 9   CPI           7605 non-null   float64
 10  Unemployment  7605 non-null   float64
 11  IsHoliday     8190 non-null   bool   
dtypes: bool(1), float64(9), int64(1), object(1)
memory usage: 712.0+ KB


In [9]:
df_features['Date']=pd.to_datetime(df_features['Date'])

In [10]:
df_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         8190 non-null   int64         
 1   Date          8190 non-null   datetime64[ns]
 2   Temperature   8190 non-null   float64       
 3   Fuel_Price    8190 non-null   float64       
 4   MarkDown1     4032 non-null   float64       
 5   MarkDown2     2921 non-null   float64       
 6   MarkDown3     3613 non-null   float64       
 7   MarkDown4     3464 non-null   float64       
 8   MarkDown5     4050 non-null   float64       
 9   CPI           7605 non-null   float64       
 10  Unemployment  7605 non-null   float64       
 11  IsHoliday     8190 non-null   bool          
dtypes: bool(1), datetime64[ns](1), float64(9), int64(1)
memory usage: 712.0 KB


In [11]:
df_features.shape

(8190, 12)

In [12]:
df_features.duplicated().sum()

np.int64(0)

In [13]:
df_features.isna().sum()

Store              0
Date               0
Temperature        0
Fuel_Price         0
MarkDown1       4158
MarkDown2       5269
MarkDown3       4577
MarkDown4       4726
MarkDown5       4140
CPI              585
Unemployment     585
IsHoliday          0
dtype: int64

In [14]:
#treating the missing values in markdown
#MarkDown1 
((df_features.isna().sum()/df_features.shape[0])*100).round(2)

Store            0.00
Date             0.00
Temperature      0.00
Fuel_Price       0.00
MarkDown1       50.77
MarkDown2       64.33
MarkDown3       55.89
MarkDown4       57.70
MarkDown5       50.55
CPI              7.14
Unemployment     7.14
IsHoliday        0.00
dtype: float64

In [15]:
##Markdown are promotional campaigns.
##Missing values are assumed no promotional events
##Hence replaced with 0
markdown_cols = ["MarkDown1", "MarkDown2", "MarkDown3", "MarkDown4", "MarkDown5"]
df_features[markdown_cols] = df_features[markdown_cols].fillna(0)

In [16]:
df_features["CPI"] = df_features["CPI"].ffill().bfill()
df_features["Unemployment"] = df_features["Unemployment"].ffill().bfill()

In [17]:
df_features.isna().sum()

Store           0
Date            0
Temperature     0
Fuel_Price      0
MarkDown1       0
MarkDown2       0
MarkDown3       0
MarkDown4       0
MarkDown5       0
CPI             0
Unemployment    0
IsHoliday       0
dtype: int64

In [18]:
df_features.dtypes

Store                    int64
Date            datetime64[ns]
Temperature            float64
Fuel_Price             float64
MarkDown1              float64
MarkDown2              float64
MarkDown3              float64
MarkDown4              float64
MarkDown5              float64
CPI                    float64
Unemployment           float64
IsHoliday                 bool
dtype: object

In [19]:
df_features.describe()

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment
count,8190.000000,8190,8190.000000,8190.000000,8190.000000,8190.000000,8190.000000,8190.000000,8190.000000,8190.000000,8190.000000
mean,23.000000,2011-10-31 12:00:00,59.356198,3.405992,3462.090725,1206.981664,776.464219,1392.763115,2043.403725,172.855800,7.747708
min,1.000000,2010-02-05 00:00:00,-7.290000,2.472000,-2781.450000,-265.760000,-179.260000,0.000000,-185.170000,126.064000,3.684000
25%,12.000000,2010-12-17 00:00:00,45.902500,3.041000,0.000000,0.000000,0.000000,0.000000,0.000000,132.653774,6.565000
50%,23.000000,2011-10-31 12:00:00,60.710000,3.513000,0.000000,0.000000,0.000000,0.000000,0.000000,182.855869,7.742000
75%,34.000000,2012-09-14 00:00:00,73.880000,3.743000,4639.585000,98.590000,24.220000,774.692500,2680.295000,214.367504,8.549000
max,45.000000,2013-07-26 00:00:00,101.950000,4.468000,103184.980000,104519.540000,149483.310000,67474.850000,771448.100000,228.976456,14.313000
std,12.987966,NaN,18.678607,0.431337,7388.916286,5495.556015,7539.953758,4707.111488,9431.223215,39.898285,1.878347


In [20]:
df_features.columns

Index(['Store', 'Date', 'Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2',
       'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment',
       'IsHoliday'],
      dtype='object')

In [21]:
#checking if store IDs are unique
df_features['Store'].nunique()

45

# Data Cleaning ## df_Stores

In [22]:
df_stores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45 entries, 0 to 44
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Store   45 non-null     int64 
 1   Type    45 non-null     object
 2   Size    45 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 1.2+ KB


In [23]:
df_stores['Type'].unique()

array(['A', 'B', 'C'], dtype=object)

In [24]:
df_stores['Type'].value_counts()

Type
A    22
B    17
C     6
Name: count, dtype: int64

In [25]:
df_stores.isna().sum()

Store    0
Type     0
Size     0
dtype: int64

In [26]:
df_stores.duplicated().sum()

np.int64(0)

In [27]:
df_stores.shape

(45, 3)

In [28]:
df_stores.describe()

,Store,Size
count,45.000000,45.000000
mean,23.000000,130287.600000
std,13.133926,63825.271991
min,1.000000,34875.000000
25%,12.000000,70713.000000
50%,23.000000,126512.000000
75%,34.000000,202307.000000
max,45.000000,219622.000000


In [29]:
#checking if store IDs are unique
df_stores['Store'].nunique()

45

# Data cleaning ## df_train

In [30]:
df_train.columns

Index(['Store', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday'], dtype='object')

In [31]:
df_train.isna().sum()

Store           0
Dept            0
Date            0
Weekly_Sales    0
IsHoliday       0
dtype: int64

In [32]:
df_train.duplicated().sum()

np.int64(0)

In [33]:
df_train.dtypes

Store             int64
Dept              int64
Date             object
Weekly_Sales    float64
IsHoliday          bool
dtype: object

In [34]:
df_train['Date']=pd.to_datetime(df_train['Date'])

In [35]:
df_train.dtypes

Store                    int64
Dept                     int64
Date            datetime64[ns]
Weekly_Sales           float64
IsHoliday                 bool
dtype: object

In [36]:
df_train.shape

(421570, 5)

In [37]:
df_train.describe()

,Store,Dept,Date,Weekly_Sales
count,421570.000000,421570.000000,421570,421570.000000
mean,22.200546,44.260317,2011-06-18 08:30:31.963375104,15981.258123
min,1.000000,1.000000,2010-02-05 00:00:00,-4988.940000
25%,11.000000,18.000000,2010-10-08 00:00:00,2079.650000
50%,22.000000,37.000000,2011-06-17 00:00:00,7612.030000
75%,33.000000,74.000000,2012-02-24 00:00:00,20205.852500
max,45.000000,99.000000,2012-10-26 00:00:00,693099.360000
std,12.785297,30.492054,NaN,22711.183519


In [38]:
#checking if store IDs are unique
df_train['Store'].nunique()

45

In [39]:
df_train.columns

Index(['Store', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday'], dtype='object')

In [40]:
#checking negative weekly sales
df_train[df_train['Weekly_Sales']<0].head()

,Store,Dept,Date,Weekly_Sales,IsHoliday
846,1,6,2012-08-10,-139.65,False
2384,1,18,2012-05-04,-1.27,False
6048,1,47,2010-02-19,-863.00,False
6049,1,47,2010-03-12,-698.00,False
6051,1,47,2010-10-08,-58.00,False


In [41]:
negative_sales=df_train[df_train['Weekly_Sales']<0].shape[0]
print('The negative sales record are', negative_sales)

The negative sales record are 1285


In [42]:
df_train.shape

(421570, 5)

In [43]:
((df_train[df_train['Weekly_Sales']<0].shape[0]/df_train.shape[0])*100)

0.30481296107408024

In [44]:
df_train[df_train['Weekly_Sales']<0].nunique()

Store            45
Dept             50
Date            143
Weekly_Sales    678
IsHoliday         2
dtype: int64

### Handling Negative Weekly Sales
There are 1285 total rows showing negative sales. out of which there are 45 unique stores, 50 different dept, 143 unique dates, 678 sales value, and 2 holidays meaning returns have been done during holiday weeks and during non holiday weeks both.The negative sales could be refunds or returns. since we are forecasting demands I would remove the negative sales as they are nott the actual demands and also these are just 0.3% of all sales


### Removing negative sales

In [45]:
df_train=df_train[df_train['Weekly_Sales']>=0]

In [46]:
len(df_train[df_train['Weekly_Sales']<0])

0

In [47]:
##Date range in train set
print('The start date is:', df_train['Date'].min())
print('The end date is:', df_train['Date'].max())

The start date is: 2010-02-05 00:00:00
The end date is: 2012-10-26 00:00:00


In [48]:
## Date range in feature set
print('The start date is:', df_features['Date'].min())
print('The end date is:', df_features['Date'].max())

The start date is: 2010-02-05 00:00:00
The end date is: 2013-07-26 00:00:00


# Merge all datasets on Date and store

In [49]:
df_features.columns

Index(['Store', 'Date', 'Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2',
       'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment',
       'IsHoliday'],
      dtype='object')

In [50]:
df_stores.columns

Index(['Store', 'Type', 'Size'], dtype='object')

In [51]:
df_train.columns

Index(['Store', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday'], dtype='object')

In [52]:
#testing whether the 'store' column is uniform in all 3 datasets
set(df_features['Store'])==set(df_stores['Store'])

True

In [53]:
set(df_stores['Store'])==set(df_train['Store'])

True

In [54]:
#testing whether the 'date' column matches in train and feature sets

In [55]:
set(df_train['Date'])==set(df_train['Date'])

True

In [56]:
#testing whether the 'isHoliday' column matches in train and feature sets

In [57]:
set(df_features['IsHoliday'])== set(df_train['IsHoliday'])

True

In [58]:
df_final=pd.merge(df_features, df_train,on=['Store', 'Date', 'IsHoliday'])
df_final=pd.merge(df_final, df_stores,on='Store')

In [59]:
df_final.head()

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday,Dept,Weekly_Sales,Type,Size
0,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,1,24924.50,A,151315
1,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,2,50605.27,A,151315
2,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,3,13740.12,A,151315
3,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,4,39954.04,A,151315
4,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,5,32229.38,A,151315


In [60]:
### Saving df_final to csv
df_final.to_csv('final_merged_data.csv', index=False)

In [61]:
pd.read_csv('final_merged_data.csv').head(2)

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday,Dept,Weekly_Sales,Type,Size
0,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,1,24924.50,A,151315
1,1,2010-02-05,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False,2,50605.27,A,151315
